metadata-ingestion/src/datahub/ingestion/source/sac/sac.py

```diff
@@ -0,0 +1,808 @@
+    )
+
+    @validator("tenant_url", "token_url")
+    def remove_trailing_slash(cls, v):
+        return config_clean.remove_trailing_slashes(v)
+
+
+@dataclass
+class SACSourceReport(StaleEntityRemovalSourceReport):
+    pass
+
+
+@platform_name("SAP Analytics Cloud", id="sac")
+@config_class(SACSourceConfig)
+@support_status(SupportStatus.TESTING)
+@capability(SourceCapability.PLATFORM_INSTANCE, "Enabled by default")
+@capability(SourceCapability.DESCRIPTIONS, "Enabled by default")
+@capability(
+    SourceCapability.LINEAGE_COARSE,
+    "Enabled by default (only for Live Data Models)",
+)
+@capability(SourceCapability.DELETION_DETECTION, "Enabled via stateful ingestion")
+@capability(
+    SourceCapability.SCHEMA_METADATA,
+    "Enabled by default (only for Import Data Models)",
+)
+class SACSource(StatefulIngestionSourceBase, TestableSource):
+    config: SACSourceConfig
+    report: SACSourceReport
+    platform = "sac"
+
+    session: OAuth2Session
+    client: pyodata.Client
+
+    ingested_dataset_entities: Set[str] = set()
+    ingested_upstream_dataset_keys: Set[str] = set()
+
+    def __init__(self, config: SACSourceConfig, ctx: PipelineContext):
+        super().__init__(config, ctx)
+        self.config = config
+        self.report = SACSourceReport()
+
+        self.session = OAuth2Session(
+            client_id=self.config.client_id,
+            client_secret=self.config.client_secret.get_secret_value(),
+            token_endpoint=config.token_url,
+            token_endpoint_auth_method="client_secret_post",
+            grant_type="client_credentials",
+        )
+
+        self.session.register_compliance_hook(
+            "protected_request", _add_sap_sac_custom_auth_header
+        )
+        self.session.fetch_token()
+
+        self.client = pyodata.Client(
+            url=f"{self.config.tenant_url}/api/v1",
+            connection=self.session,
+            config=pyodata.v2.model.Config(retain_null=True),
+        )
+
+    def close(self) -> None:
+        self.session.close()
+        super().close()
+
+    @classmethod
+    def create(cls, config_dict: dict, ctx: PipelineContext) -> "SACSource":
+        config = SACSourceConfig.parse_obj(config_dict)
+        return cls(config, ctx)
+
+    @staticmethod
+    def test_connection(config_dict: dict) -> TestConnectionReport:
+        test_report = TestConnectionReport()
+
+        try:
+            config = SACSourceConfig.parse_obj(config_dict)
+
+            session = OAuth2Session(
+                client_id=config.client_id,
+                client_secret=config.client_secret.get_secret_value(),
+                token_endpoint=config.token_url,
+                token_endpoint_auth_method="client_secret_post",
+                grant_type="client_credentials",
+            )
+
+            session.register_compliance_hook(
```

Comment:
is there a way to extract more of this connection logic out into a helper? don't like that it's repeated across init and test_connection

e.g. `self.session, self.client = make_sac_connection(config)`